tests/Makefile

```diff
@@ -57,6 +57,7 @@ else
 DLEXT ?= so
 EXEXT=
 SHARED_SUFFIX=
+LOCAL_LDFLAGS += -lpthread
```

Comment:
CMake does not have the issue as it is searching for the PThreads package, and will link it if it finds it.

I *hope* that this will be fixed if we remove the naked PThreads calls from the tests via an abstraction, as we should no longer need to explicitly link it. This is a problem that only occurs with older compilers (unfortunately by this I also include Ubuntu 22.04) as libpthread has been rolled into glibc, and thus explicit linking is no longer required  see: https://developers.redhat.com/articles/2021/12/17/why-glibc-234-removed-libpthread

Trying to detect compile options and only link in Make if its set is not going to be simple, to put it mildly. `ssl_pthread_server` is generally not built unless you pass `PTHREAD` on the command line. The tests as they stand at the minute will not link without explicit pthread linking, and not building the tests is not an option. 

I will attempt the abstraction, and report back.

Reply:
The problem is what happens on a system that doesn't have pthreads. This must not prevent building the library in configurations where pthreads isn't needed.

Reply:
Yes, so when we want to build pthreads we have to call `make tests TEST_THREADING=pthreads`, i.e. we only do this when we know we're running on such a system.

The shell could use `uname -s` to determine this - the `Makefile` could too, but since `Makefile`s might be run on Windows, it might be better to have the invoking shell do this

Reply:
If the compiler is GCC-like, we can run something like `gcc -E $(CFLAGS) $(LOCAL_CFLAGS) scripts/data_files/pthread_check.c` where `pthread_check.c` contains
```
#include <mbedtls/build_info.h>
#if MBEDTLS_THREADING_PTHREAD
mbedtls_must_be_linked_with_pthread
#endif
```
and see if the output includes `mbedtls_must_be_linked_with_pthread`. Making this work with other compilers would be more complicated. Maybe arrange for `mbedtls_must_be_linked_with_pthread` to be a symbol, do a normal compilation with `-c` and search the symbol in the object file? The search would need to assume either a non-Windows system with `grep` or a Windows system with [`find`](https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/find) (if that works on a binary file, which I'm not sure.)

Reply:
`cl.exe` has the `/P`option, but it outputs to a `.i` file with the same basename as the input `.c` file.

But that's why I think this is best done in the invoking shell, rather than the Makefile - the shell can test the environment it's on (and TBH on Windows there's no specific library linkage - `CreateThread()` has been around since WIndows XP, and is in `kernel32.lib`; `AcquireSRWLockExclusive()` and friends have been around since Windows Vista (urgh) and are also in `kernel32.lib`)

Reply:
To be clear, I think it's perfectly fine if users who enable `MBEDTLS_THREADING_PTHREAD` need to pass an extra `LDFLAGS`. This is already the case today except in the corner case of enabling the option but not calling any library function that has a lock or otherwise using threads in the application.

What I objected to in this thread was linking with pthread unconditionally, which breaks the build for anyone whose platform or compiler doesn't support `-lpthread`. For example MSVC — our makefile is supposed to work with MSVC (although I don't think we test this on the CI).

Reply:
Done, I think.